src/shared/awsagent/roles.go

```diff
@@ -173,48 +175,185 @@ func (a *Agent) CreateRolesAnywhereProfileForRole(ctx context.Context, role type
-			_, err := a.iamClient.UntagRole(ctx, &iam.UntagRoleInput{RoleName: role.RoleName, TagKeys: []string{softDeletedTagKey}})
-			if err != nil {
-				return nil, errors.Wrap(err)
+	var existingPolicy map[string]any
+	err = json.Unmarshal([]byte(assumeRolePolicyDocumentRaw), &existingPolicy)
+	if err != nil {
+		return false, errors.Wrap(err)
+	}
+
+	// For some reason, Action can either be a list or a string.
+	statements := existingPolicy["Statement"].([]any)
+	for i, statement := range statements {
+		statementTyped := statement.(map[string]any)
+		// Get action or actions
+		if action, ok := statementTyped["Action"]; ok {
+			switch actionType := action.(type) {
+			case string:
+				statementTyped["Action"] = []string{actionType} // Convert single string to slice
+				statements[i] = statementTyped                  // Update the statement in the slice
+			case []any:
+				// do nothing
+			default:
+				return false, errors.Errorf("unexpected type for Action: %T", actionType)
 			}
-			// Intentionally not returning here, as we want to continue and check if we need to mark as soft delete only
 		}
-		if useSoftDeleteStrategy {
-			err = a.setRoleSoftDeleteStrategyTag(ctx, role)
-			if err != nil {
-				return nil, errors.Wrap(err)
+	}
+	existingPolicy["Statement"] = statements
+
+	// Marshal the policy back to JSON
+	assumeRolePolicyDocument, err := json.Marshal(existingPolicy)
+	if err != nil {
+		return false, errors.Wrap(err)
+	}
+
+	var existingPolicyStructured PolicyDocument
+	err = json.Unmarshal(assumeRolePolicyDocument, &existingPolicyStructured)
+	if err != nil {
+		return false, errors.Wrap(err)
+	}
+
+	if existingPolicyStructured.Version != newDocument.Version {
+		return false, nil
+	}
+
+	if len(existingPolicyStructured.Statement) != len(newDocument.Statement) {
+		return false, nil
+	}
+
+	for _, newStatement := range newDocument.Statement {
+		found := false
+		for _, existingStatement := range existingPolicyStructured.Statement {
+			// Fields: Effect, Action, Resource, Principal, Sid, Condition
+			// Action is a slice
+			// Principal is a map
+			// Condition is a map
+			if newStatement.Effect != existingStatement.Effect {
+				continue
 			}
-		}
-		if !useSoftDeleteStrategy && HasSoftDeleteStrategyTagSet(role.Tags) {
-			err = a.UnsetRoleSoftDeleteStrategyTag(ctx, role)
-			if err != nil {
-				return nil, errors.Wrap(err)
+			slices.Sort(existingStatement.Action)
+			slices.Sort(newStatement.Action)
+			if !slices.Equal(existingStatement.Action, newStatement.Action) {
+				continue
+			}
+			if newStatement.Resource != existingStatement.Resource {
+				continue
+			}
+			existingPrincipals := lo.MapToSlice(existingStatement.Principal, func(key string, value string) string {
+				return fmt.Sprintf("%s:%s", key, value)
+			})
+			newPrincipals := lo.MapToSlice(newStatement.Principal, func(key string, value string) string {
+				return fmt.Sprintf("%s:%s", key, value)
+			})
+			slices.Sort(existingPrincipals)
```

Comment:
can we sort elements on existing statements once before the for loop? rather than over and over again on every newStatement iteration.